### PR TITLE
Improvements to the Worker API

### PR DESCRIPTION
**Pull request by ghale**
This introduces changes to the Worker API to make it more coherent and easier to use.  The changes involve separating the worker requirements from the parameters for a given item of work.  It also introduces a typed configuration object for the worker requirements that only allows you to configure the requirements appropriate for the type of worker you're using.  Typical usage would look like:

### Workers with no isolation
```
// Get a work queue for work without isolation
WorkQueue queue = workerExecutor.noIsolation()

// OR

// Get a work queue for work with classloader isolation
WorkQueue queue = workerExecutor.classLoaderIsolation() {
    classpath.from(configurations.extraClasspath)
}

// OR

// Get a work queue for work with process isolation
WorkQueue queue = workerExecutor.processIsolation() {
    classpath.from(configurations.extraClasspath)
    forkOptions {
        maxHeapSize = "128m"
        systemProperty("foo", "bar")
    }
}

// Then submit as many items to the queue as you want, specifying the parameters for each item of work
inputFiles.each { File inputFile ->
    queue.submit(TestWorkAction.class) { TestWorkParameters ->
        fileParameter = inputFile
    }
}

// Optionally wait for all work submitted to this queue
queue.await()
```

This also renames the recently introduced `WorkerExecution` and `WorkerParameters` classes to `WorkAction` and `WorkParameters`, respectively.

### CODE REVIEW

File: subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorLegacyApiIntegrationTest.groovy [hunk 1]
```diff
@@ -186,4 +275,40 @@ class WorkerExecutorLegacyApiIntegrationTest extends AbstractIntegrationSpec {
...
+
+                public void run() {
+                    RuntimeMXBean runtimeMxBean = ManagementFactory.getRuntimeMXBean();
+                    List<String> arguments = runtimeMxBean.getInputArguments();
+                    assert arguments.contains("-Dfoo=bar");
+                    assert arguments.contains("-Dbar=baz");
+                    assert arguments.contains("-Xmx128m");
+                    assert arguments.contains("-Xms128m");
+                    assert arguments.contains("-Dfile.encoding=UTF-8");
+                    assert arguments.contains("-ea");
```
Comment (lacasseio) [inline subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorLegacyApiIntegrationTest.groovy]: ❓ Nit: Would there be a way to centralize those configurations so it's declared once but used in multiple places?

File: subprojects/workers/src/main/java/org/gradle/workers/ProcessWorkerSpec.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.workers;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A worker spec providing the requirements of a forked process with a custom classpath.
```
Comment (lacasseio) [inline subprojects/workers/src/main/java/org/gradle/workers/ProcessWorkerSpec.java]: ❓ Here we assume a forked JVM process, could it be just a normal process?

File: subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorLegacyApiIntegrationTest.groovy (hunk 1 above)
Comment (ghale) [reply subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorLegacyApiIntegrationTest.groovy]: Yeah, this has always bothered me.  I made it a little tighter by using a "verifier" object that produces the dsl to configure it, captures the process options as a json file, and then verifies the options match.

Comment (ghale) [reply subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorLegacyApiIntegrationTest.groovy]: See https://github.com/gradle/gradle/pull/10047/commits/3d06528e6994b508a0300f0b873f6da89496fa14

**Pull request merged**